spring-xd-messagebus-rabbit/src/main/java/org/springframework/xd/dirt/integration/rabbit/RabbitMessageBus.java

```diff
@@ -673,6 +682,31 @@ public void destroy() {
 		stopBindings();
 	}
 
+	@Override
+	public void doManualAck(LinkedList<MessageHeaders> messageHeadersList) {
+		Iterator<MessageHeaders> iterator = messageHeadersList.iterator();
+		Map<Object, Long> channelsToAck = new HashMap<Object, Long>();
+		while(iterator.hasNext()) {
+			MessageHeaders messageHeaders = iterator.next();
+			if (messageHeaders.containsKey(AmqpHeaders.CHANNEL)) {
+				Channel channel = (com.rabbitmq.client.Channel) messageHeaders.get(AmqpHeaders.CHANNEL);
+				Long deliveryTag = (Long) messageHeaders.get(AmqpHeaders.DELIVERY_TAG);
+				// todo: If the order of messages isn't preserved in spite of storing the incoming
+				// todo: messages into LinkedList, how do we handle the deliveryTag?
+				channelsToAck.put(channel, deliveryTag);
```

Comment:
The delivery tags (by channel) should be in the right (delivery) order so I believe it's safe to (multiple) ack the last tag for each channel.

Understand, though, that we might get duplicates messages. Connection resets half-way through a block, the already unacked messages will be redelivered and added to the same block.

But the delivery tag handling will be correct.